### PR TITLE
Add trackError API

### PR DESCRIPTION
**Pull request by AnastasiaKubova**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [ ] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

- Add `trackError` API.
- Update tests.

## Related PRs or issues

[AB#87138](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/87138)

### CODE REVIEW

File: AppCenterCrashes/AppCenterCrashes/Model/MSACException.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Creates an instance of exception model.
+ *
+ * @param exceptionType exception type.
+ * @param exceptionMessage exception message.
+ *
+ * @return A new instance of exception model.
+ */
+- (instancetype)initWithTypeAndMessage:(NSString *)exceptionType exceptionMessage:(NSString *)exceptionMessage;
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACException.h:25]: initWithType:exceptionMessage

File: AppCenterCrashes/AppCenterCrashes/Model/MSACException.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+#if __has_include(<AppCenter/MSACSerializableObject.h>)
+#import <AppCenter/MSACSerializableObject.h>
+#else
+#import "MSACSerializableObject.h"
+#endif
+
+@class MSACStackFrame;
+
+NS_SWIFT_NAME(ExceptionModel)
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACException.h:14]: ExceptionModel -> Exception

File: AppCenterCrashes/AppCenterCrashes/Model/MSACException.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+@property(nonatomic) NSArray<MSACStackFrame *> *frames;
+
+/**
+ * Convert NSError to MSACException.
+ *
+ * @param error - NSError object.
+ *
+ * @return MSACException exception.
+ */
++ (MSACException *)convertNSErrorToMSACException:(NSError *)error;
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACException.h:63]: why not initWithError:(NSError)?

File: AppCenterCrashes/AppCenterCrashes/Model/MSACException.m [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      self.message = exception.reason;
+    }
+    if ([exception respondsToSelector:NSSelectorFromString(@"callStackSymbols")]) {
+      self.stackTrace = exception.callStackSymbols.description;
+    }
+  }
+  return self;
+}
+
+- (instancetype)initWithTypeAndMessage:(NSString *)exceptionType exceptionMessage:(NSString *)exceptionMessage {
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACException.m:32]: initWthType:exceptionMessage

File: AppCenterCrashes/AppCenterCrashes/MSACCrashes.h [hunk 5]
```diff
@@ -69,6 +71,32 @@ typedef NS_ENUM(NSUInteger, MSACUserConfirmation) {
...
+/**
+ * Track handled exception from custom exception model.
+ *
+ * @param exceptionModel custom model exception.
+ * @param properties dictionary of properties.
+ * @param attachments a list of attachments.
+ *
+ * @return handled error ID.
+ */
++ (NSString *_Nonnull)trackException:(MSACException *_Nonnull)exceptionModel
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/MSACCrashes.h:96]: exceptionModel -> exception
Also fix the method doc pls

File: AppCenterCrashes/AppCenterCrashes/MSACCrashes.h [hunk 6]
```diff
@@ -69,6 +71,32 @@ typedef NS_ENUM(NSUInteger, MSACUserConfirmation) {
...
+/**
+ * Track handled error.
+ *
+ * @param error error.
+ * @param properties dictionary of properties.
+ * @param attachments a list of attachments.
+ *
+ * @return handled error ID.
+ */
++ (NSString *_Nonnull)trackError:(NSError *_Nonnull)error
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/MSACCrashes.h:83]: We should probably swiftify these methods, like:
```objc
NS_SWIFT_NAME(trackError(_:properties:attachments)
```

File: AppCenterCrashes/AppCenterCrashes/MSACCrashes.mm [hunk 7]
```diff
@@ -208,6 +209,38 @@ @implementation MSACCrashes
 
 #pragma mark - Public Methods
 
+/**
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/MSACCrashes.mm:212]: API docs are useless in `.mm` files, let's remove it.

File: AppCenterCrashes/AppCenterCrashes/MSACCrashes.mm [hunk 8]
```diff
@@ -1354,9 +1388,9 @@ + (void)resetSharedInstance {
 
 #pragma mark - Handled exceptions
 
-- (NSString *)trackModelException:(MSACException *)exception
-                   withProperties:(nullable NSDictionary<NSString *, NSString *> *)properties
-                  withAttachments:(nullable NSArray<MSACErrorAttachmentLog *> *)attachments {
+- (NSString *)trackError:(MSACExceptionModel *_Nonnull)exception
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/MSACCrashes.mm:1391]: `_Nonnull` is not needed here, and `nullable` too.

File: AppCenterCrashes/AppCenterCrashes/Model/MSACExceptionModel.h [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+@interface MSACExceptionModel : NSObject <MSACSerializableObject>
+
+/**
+ * Creates an instance of exception model.
+ *
+ * @param error error.
+ *
+ * @return A new instance of exception.
+ */
+- (instancetype)initWithError:(NSError *)error;
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACExceptionModel.h:27]:
```objc
NS_SWIFT_NAME(init(withError:))
```

File: AppCenterCrashes/AppCenterCrashes/Model/MSACExceptionModel.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Creates an instance of exception model.
+ *
+ * @param exceptionType exception type.
+ * @param exceptionMessage exception message.
+ * @param stackTrace stack trace.
+ *
+ * @return A new instance of exception.
+ */
+- (instancetype)initWithType:(NSString *)exceptionType
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACExceptionModel.h:38]:
```objc
NS_SWIFT_NAME(init(withType:exceptionMessage:stackTrace))
```

File: AppCenterCrashes/AppCenterCrashes/Model/MSACExceptionModel.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+                  stackTrace:(NSArray<NSString *> *)stackTrace;
+
+/**
+ * Creates an instance of exception model.
+ *
+ * @exception exception.
+ *
+ * @return A new instance of exception.
+ */
+- (instancetype)initWithException:(NSException *)exception;
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACExceptionModel.h:49]:
```objc
NS_SWIFT_NAME(init(withException:))
```

File: AppCenterCrashes/AppCenterCrashes/Model/MSACWrapperExceptionModel.h [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+#if __has_include(<AppCenterCrashes/MSACExceptionModel.h>)
+#import <AppCenterCrashes/MSACExceptionModel.h>
+#import <AppCenterCrashes/MSACWrapperExceptionModel.h>
+#else
+#import "MSACExceptionModel.h"
+#import "MSACWrapperExceptionModel.h"
+#endif
+
+#if __has_include(<AppCenter/MSACSerializableObject.h>)
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACWrapperExceptionModel.h:17]: One `#if` should be enough. Just put these imports in first condition.

File: AppCenterCrashes/AppCenterCrashes/Model/MSACWrapperExceptionModel.h [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+@interface MSACWrapperExceptionModel : MSACExceptionModel
+
+/*
+ * Inner exceptions of this exception [optional].
+ */
+@property(nonatomic) NSArray<MSACWrapperExceptionModel *> *innerExceptions;
+
+/*
+ * Name of the wrapper SDK that emitted this exception.
+ * Consists of the name of the SDK and the wrapper platform, e.g. "appcenter.xamarin", "hockeysdk.cordova" [optional].
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Model/MSACWrapperExceptionModel.h:32]:
```suggestion
 * Consists of the name of the SDK and the wrapper platform, e.g. "appcenter.xamarin", "appcenter.react-native" [optional].
```

File: AppCenterCrashes/AppCenterCrashes/WrapperSDKUtilities/MSACWrapperCrashesHelper.h [hunk 14]
```diff
@@ -13,7 +13,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 @class MSACErrorReport;
 @class MSACErrorAttachmentLog;
-@class MSACException;
+@class MSACExceptionModel;
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/WrapperSDKUtilities/MSACWrapperCrashesHelper.h:16]: I think we can remove this line now.

File: AppCenterCrashes/AppCenterCrashesTests/MSACStackFrameTests.m [hunk 15]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#import "AppCenter+Internal.h"
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashesTests/MSACStackFrameTests.m:4]: do we need this line?

File: AppCenterCrashes/AppCenterCrashes/Model/MSACWrapperExceptionModel.h (hunk 12 above)
Comment (AnastasiaKubova) [reply AppCenterCrashes/AppCenterCrashes/Model/MSACWrapperExceptionModel.h:17]: no, these imports from different modules (core and crashes)

File: AppCenterCrashes/AppCenterCrashesTests/MSACStackFrameTests.m (hunk 15 above)
Comment (AnastasiaKubova) [reply AppCenterCrashes/AppCenterCrashesTests/MSACStackFrameTests.m:4]: yes, it is needed for using some methods (for example from `MSACUtility` class)

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,11 +1,17 @@
 # App Center SDK for iOS, macOS and tvOS Change Log
 
-## Version 4.2.1
+## Version 4.2.1 (Under development)
+
+### App Center Crashes
+
+* **[Feature]** Add support for handled errors with `Crashes.trackError` and `Crashes.trackException` APIs.
```
Comment (DmitriyKirakosyan) [inline CHANGELOG.md:7]:
```suggestion
* **[Feature]** Add support for tracking handled errors with `Crashes.trackError` and `Crashes.trackException` APIs.
```

**Review by DmitriyKirakosyan — APPROVED** [review]

**Comment by DmitriyKirakosyan** [comment]
.

**Comment by DmitriyKirakosyan** [comment]
.

**Comment by DmitriyKirakosyan** [comment]
.

**Comment by DmitriyKirakosyan** [comment]
.

**Comment by DmitriyKirakosyan** [comment]
.

**Comment by DmitriyKirakosyan** [comment]
.

**Pull request merged**